### ISSUE TITLE
Both naming checks enumerate AST/export shapes from memory — do one systematic pass against the grammar

### ISSUE THREAD

**Issue by zernie** [zernie/agent-lint#170]
Follow-up to #169, deliberately **not** fixed there. One open finding is carried here, plus the pattern behind it, which is the more useful half.

## The open finding (verified, reproduced)

`scripts/check-internal-tag.mjs` misses a declaration exported through a specifier:

```ts
/** @internal */
const widget = () => {};
export { widget };
```

`VALUE_DECL` requires an inline `export` modifier, so the declaration is never recorded — while the api report still lists `widget` as public. Result:

```
public @experimental/@internal declarations checked: 0  (across 1 api reports)
findings: 0
```

Exit 0. The contradiction the script exists to catch is invisible, and `checked: 0` is the tell.

The fix shape: collect tagged value declarations regardless of inline export syntax, then use the api report to decide whether they are public. That is the same correction already applied to the ESLint rule in #169 — which is exactly why this belongs in a pass rather than a patch.

## Why a pass and not a patch

Seventeen review findings landed on #169. **Nine were in `eslint-rules/experimental-name.mjs` alone**, and they arrived in sequence, each after the previous fix was called complete:

| round | what was unnamed |
| --- | --- |
| 1 | `export { experimental_x as x }` — aliased re-export |
| 2 | `/** @experimental */ const x = …; export { x }` — tagged local, separately exported |
| 3 | `export default function x() {}` |
| 4 | `export const { x } = source` — destructured binding |

Plus `export default` independently missing from `check-internal-tag.mjs`, and now the separately-exported case there too. **The same construct went unnamed in two different checkers written days apart**, which is not a typo repeated — it is a blind spot in how "exported" gets enumerated.

The common cause is that both checks enumerate shapes **from memory** instead of from the language. Patch-on-report converges slowly and never tells you when it is done.

## What the pass should produce

1. **Enumerate against the ESTree grammar**, not recollection: every node type that can introduce an exported binding, listed explicitly, with the ones deliberately excluded named and justified. #169 started this — the rule's header now lists its three export paths — but the list was assembled reactively and has no claim to completeness.
2. **One shared "what does this file export, and under what name" helper**, used by both the ESLint rule and the script. Today they answer the same question with two independent regexes/matchers and disagree.
3. **A fixture corpus** of every export form, run against both. That turns "did we cover it" from a memory question into a test.

## Two constraints the pass must respect, learned the hard way in #169

- **For a rule wired at `error`, a false positive is worse than a miss.** A miss costs a detection; a false positive costs the build, and a rule that fails valid code gets switched off rather than fixed. #169 shipped one (a tagged local nested in a function body condemned an unrelated module-level export of the same name) and it must not recur. Err toward silence.
- **`sourceCode.getScope(node)` is not the binding's scope.** For a `FunctionDeclaration` it returns the scope the function *creates*. The first fix for the false positive used it and broke the hoisting case; the correct test was the declaration's parent being the `Program`.

## Known limits worth keeping stated rather than silently unhandled

- An **anonymous default** (`export default () => {}`) has no name to carry a marker — a limit of a name-based convention, not a bug.
- A `var`/function inside a **top-level block** hoists to module scope and is not tracked. Accepted, because it errs toward silence.

**Comment by zernie** [comment]
Done in #186 (`94af23a5`) — the pass, not a fifth patch.

**The enumeration exists.** `scripts/lib/export-forms.mjs`: 26 rows, 22 forms both checks must handle and 4 deliberate exclusions each carrying its reason, because an exclusion nobody wrote down is indistinguishable from an oversight. All four forms that arrived one review round at a time are named explicitly, so deleting one is a visible act.

**The script stopped enumerating from memory.** `check-internal-tag.mjs` now asks the TypeScript compiler (`scripts/lib/tagged-declarations.mjs`) instead of `VALUE_DECL`. Same reflex this repo already applies to Bash and markdown, turned on its own tooling; `typescript` was already a direct dependency, so it is adopt rather than build.

**A fixture corpus drives both.** `scripts/lib/export-forms.test.ts` asserts the parser against every covered row, and `eslint-rules/experimental-name.corpus.test.ts` runs the rule over the same list in both directions — every form silent when correctly prefixed, every form reported when not. "Did we cover it" is a test now.

## What the pass found that four rounds of review had not

**1. The open finding, closed.** `/** @internal */ const widget = …; export { widget }` went from `checked: 0, findings: 0, exit 0` to correctly reported.

**2. The script had never once looked at a TYPE.** `VALUE_DECL` matched `function|const|let|var|class`; `interface`, `type` and `enum` were invisible. The moment the parser could see them, 13 tagged types appeared, **7 of them exported and `@internal`** — `EffectRegion`, `NeedsContract`, `PipeStep`, `Supplies`, `Handoff`, `Pipeline`, `KnownAgentName`.

All seven are deliberate: typed-composition machinery exported **only so a consumer's inference resolves**. Nobody writes those names, `@internal` is the truthful way to say so, and failing the build over them would force a meaningless rename or dropping a true tag. So types are now excluded **as a stated decision**, matching `local/experimental-name` (the convention is about call sites; a type annotation is not one) — rather than the accident it had been. The skip is counted and printed, because this file's own history is a counter that counted a symbol it then ignored and the number read as coverage:

```
public @experimental/@internal VALUE declarations checked: 19  (across 10 api reports)
types skipped (out of scope, see the comment above): 13
findings: 0
```

**3. A live gap in the ESLint rule — a fifth form.** `/** @experimental */ const widget = 1; export default widget;` reported nothing. It is the *same shape* the rule learned for `export { x }` in round 2 — a tagged local exported by a later statement — and it went unnamed anyway. Fixed by recording it like a specifier and judging at `Program:exit`, and it is the sharpest evidence for this issue's thesis: the list was never going to converge by report.

**4. `export { widget as default }` moved to the exclusions.** `default` is not a name any consumer writes — they choose their own at the import — so no prefix reaches a call site through it. The anonymous-default limit, reached by different syntax. The parser still *resolves* the binding (it stays covered for enumeration); only judging the NAME is out of scope, and the corpus test pins that distinction so the two do not get conflated later.

## Both constraints respected

**Err toward silence.** The rule-corpus suite asserts every form is silent when correctly named before asserting any of them fire; a false positive on a valid export would fail the valid half first.

**`getScope` is not the binding's scope.** Untouched — the `node.parent?.type === "Program"` test from #169 still stands.

## Honest about what did not get built

Item 2 of "what the pass should produce" — *one shared helper used by both the rule and the script* — is **not** done, and I would rather say so than let the closure imply it. The rule receives ESTree nodes from ESLint and must report on them; the script parses text itself with the TS compiler. Unifying those means one AST, which is a larger change than this issue's scope. What is shared is the **enumeration and the corpus**, which is what actually stops the drift: the two checks can still disagree in mechanism, but they can no longer disagree about *which forms exist* without a test failing.

Four mutations, each failing on its own assertion: the rule forgetting `export default <identifier>`, the parser dropping destructuring, the parser ignoring export specifiers, and an exclusion losing its reason.

---
_Generated by [Claude Code](https://claude.ai/code)_
